GraphBase/src/graph/DfsIterator.java

```diff
@@ -0,0 +1,63 @@
+package graph;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.Stack;
+
+public class DfsIterator implements Iterator {
+  Map<Integer,Vertex> vertices;
+  Vertex vertex;
+  int startId;
+  private HashSet<Integer> visited;
+  private Stack<Integer> s;
+  public DfsIterator(Graph graph) {
+    this.vertices = graph.getVerticesMap();
+    vertex = null;
+    visited = new HashSet<>();
+    s = new Stack<>();
+    startId = (Integer) Collections.min(graph.keySet());
+  }
+  public DfsIterator(Graph graph, int id) {
+    this.vertices = graph.getVerticesMap();
+    vertex = null;
+    visited = new HashSet<>();
+    s = new Stack<>();
+    startId = id;
+  }
+
+  public Vertex next() {
```

Comment:
The logic is very complicated, which can easily introduce bugs. I would recommend maintain one unvisited set and one stack. Every time when the stack is not empty, pop one value out and insert all its unvisited neighbours into the stack. Otherwise, take one value out of unvisited set.

Reply:
Further, it will be good if you can try make these two Iterators into one class, by defining the queue as Collection, and use Stack for dfs and LinkedList for bfs. Below is for your reference:

public class Test {
  
    public Collection<Integer> queue;
    
    public Test(boolean isBFS) {
      if (isBFS) {
        queue = new LinkedList<Integer>();
      } else {
        queue = new Stack<Integer>();
      }
    }
	
	public static void main(String[] args){
		Test test1 = new Test(true);
		Queue<Integer> q = (Queue<Integer>) test1.queue;
		q.offer(1);
		q.offer(2);
		q.poll();
		q.forEach(e -> {System.out.println(e);});
		
		Test test2 = new Test(false);
		Stack<Integer> s = (Stack<Integer>) test2.queue;
		s.push(1);
		s.push(2);
		s.pop();
		s.forEach(e-> {System.out.println(e);});	
	}
}